mgr/dashboard: Add User Info component and navigation

tspmelo opened the pull request:
Add a new entry on the navigation that shows the username of the current user.
When clicked it will open a page where the user can change its username or
password.

![image](https://user-images.githubusercontent.com/399326/38608302-835f1d64-3d6a-11e8-8512-b7081ee78bdb.png)

![image](https://user-images.githubusercontent.com/399326/38608336-97723052-3d6a-11e8-93b2-574850046808.png)

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,33 @@
+
+@ApiController('user')
+@AuthRequired()
+class User(RESTController):
```
interesting, you don't need to list users?

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,33 @@
+            'username': config_username
+        }
+
+    def set(self, data, username):
```
please use `@RESTController.args_from_json()`

capri1989 commented:
Would you mind adding a screenshot for fron tend visible changes? This makes it easier for everyone to get a first impression without building.

tspmelo commented:
@capri1989 updated PR description with screenshots.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/core/user-info/user-info.component.ts:
```diff
@@ -0,0 +1,97 @@
+  ngOnInit() {
+    this.username = this.authStorageService.get();
+    this.userService.get(this.username).subscribe((res: User) => {
+      this.userForm.patchValue(res);
```
`setValue` requires that the object you pass has values for all the `FormControls`.
In this case `res` only has the value for `username,` so it returns an exception if I try to use `setValue`.

With `patchValue` you can pass the value of any number of `FormControls` and it will update just those.

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/core/user-info/user-info.component.ts:
```diff
@@ -0,0 +1,97 @@
+    });
+  }
+
+  MatchPassword(control: AbstractControl) {
```
This does not validate the password, this is just to match the value of `password` with `confirmPassword`.

ATM I didn't add any validation for the password.
When we decide on the rules for password we should create a separate validator with the name you suggested.

votdev replied:
OK, understand. In this case i suggest to use the name ``matchPasswordValidator``. If you don't like it, you should however rename it to ``matchPassword`` (start lower case).

tspmelo commented:
Pushed the following changes:
- It's now using NotificationService and CdValidators
- added `@RESTController.args_from_json()`, as suggested by @sebastian-philipp 
- implemented the changes/fixes suggested by @votdev

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/shared/validators/cd-validators.ts:
```diff
@@ -62,4 +62,18 @@ export class CdValidators {
   }
+
+  /**
+   * Checks if the value of password matches confirmPassword.
```
Please add some information that the form controls must be named ``password`` and ``confirmPassword`` or enhance the validator to allow the developer to configure the control identifiers/paths.

tspmelo commented:
jenkins retest this please

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,30 @@
+    def set(self, username, password):
+        mgr.set_config('username', username)
+
+        if password != '':
```
If `password` is optional, the you could also do: `def set(self, username, password=None):` and `if password:`

tspmelo commented:
jenkins retest this please

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/core/user-info/user-info.component.ts:
```diff
@@ -0,0 +1,91 @@
+      },
+      null,
+      () => {
+        this.userForm.setErrors({ cdSubmitButton: true });
```
The error handling method is set for the ``complete`` instead of the ``error`` function parameter. Should be in line 81.

votdev replied:
This comment can be ignored, all's fine :-)

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/core/user-info/user-info.component.ts:
```diff
@@ -0,0 +1,91 @@
+    private notificationService: NotificationService,
+    private authService: AuthService,
+    private router: Router,
+    private ngZone: NgZone
```
Can be removed because it is not used.

tspmelo replied:
Done

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/core/user-info/user-info.component.ts:
```diff
@@ -0,0 +1,91 @@
+import { Location } from '@angular/common';
+import { Component, NgZone, OnInit } from '@angular/core';
+import {
+  AbstractControl,
```
Can be removed because it is not used.

tspmelo replied:
Done

votdev commented on src/pybind/mgr/dashboard/frontend/src/app/core/user-info/user-info.component.ts:
```diff
@@ -0,0 +1,91 @@
+  AbstractControl,
+  FormControl,
+  FormGroup,
+  ValidationErrors,
```
Can be removed because it is not used.

tspmelo replied:
Done

rjfd commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,30 @@
+class User(RESTController):
+    # pylint: disable=unused-argument
+    def get(self, username):
+        config_username = mgr.get_config('username', None)
```
I think the user management code here and in `controllers/auth.py` should be moved into a central place like the `services` directory, so that in the future if we change the key of the config, where we store the username and password, we don't need to look for all the places where it is used.

rjfd commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,30 @@
+        mgr.set_config('username', username)
+
+        if password:
+            hashed_passwd = Auth.password_hash(password)
```
If we create a user management service (as stated in the previous comment) we can put this logic hidden inside the service and just expose a well defined public interface to update the username and password without revealing the implementation details.

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/user.py:
```diff
@@ -0,0 +1,30 @@
+            'username': config_username
+        }
+
+    @RESTController.args_from_json
```
This PR needs to be rebased. `args_from_json` is gone now.

sebastian-philipp commented:
@tspmelo Can you give us an update on this PR?

tspmelo commented:
This will be included in a future PR.
